python_modules/dagster/dagster/_core/definitions/definitions_class.py

```diff
@@ -85,6 +95,69 @@ def create_repository_using_definitions_args(
     )
 
 
+class _AttachedObjects(NamedTuple):
+    jobs: Iterable[Union[JobDefinition, UnresolvedAssetJobDefinition]]
+    schedules: Iterable[Union[ScheduleDefinition, UnresolvedPartitionedAssetScheduleDefinition]]
+    sensors: Iterable[SensorDefinition]
+
+
+def _attach_resources_to_jobs_and_instigators(
+    jobs: Optional[Iterable[Union[JobDefinition, UnresolvedAssetJobDefinition]]],
+    schedules: Optional[
+        Iterable[Union[ScheduleDefinition, UnresolvedPartitionedAssetScheduleDefinition]]
+    ],
+    sensors: Optional[Iterable[SensorDefinition]],
+    resource_defs: Mapping[str, Any],
+) -> _AttachedObjects:
+    """
+    Given a list of jobs, schedules, and sensors along with top-level resource definitions,
+    attach the resource definitions to the jobs, schedules, and sensors which require them.
+    """
+    jobs = jobs or []
+    schedules = schedules or []
+    sensors = sensors or []
+
+    # Find unsatisfied jobs
+    unsatisfied_jobs = [
+        job
+        for job in jobs
+        if isinstance(job, JobDefinition) and job.is_missing_required_resources()
+    ]
+
+    # Create a mapping of job id to a version of the job with the resource defs bound
+    unsatisfied_job_to_resource_bound_job = {
+        id(job): job.with_top_level_resources(resource_defs)
```

Comment:
i like this honestly - it's direct about the fact that we use reference equality for jobs.